### PR TITLE
SB-13803 Validation for UserFreeUpController

### PR DESCRIPTION
**Pull request by anmol2302**
<b>Ticket Ref: </b>[SC-13083](https://project-sunbird.atlassian.net/browse/SB-13803)
<b>Docs: </b>Free Up user [doc](https://docs.google.com/document/d/1c94aM6g-gCuBadyBz-GnUtdGA38rC0Mv4jWb4aiZXHs/edit)
1: A new Validator class  is been added to validate the  request for UserFreeUpController.
2: it will validate the following
             - id presense
             - identifier(must be List and contains value either email,phone or both) otherwise
             - will throw exception client error(400)

### CODE REVIEW

File: sunbird-platform-core/common-util/src/main/java/org/sunbird/common/models/util/ActorOperations.java [hunk 1]
```diff
@@ -156,7 +156,8 @@
   GET_USER_TYPES("getUserTypes"),
   CLEAR_CACHE("clearCache"),
   USER_TENANT_MIGRATE("userTenantMigrate"),
-  GET_PARTICIPANTS("getParticipants");
+  GET_PARTICIPANTS("getParticipants"),
+  FREEUP_USER("freeUpUser");
```
Comment (manzarul) [inline sunbird-platform-core/common-util/src/main/java/org/sunbird/common/models/util/ActorOperations.java]: FREEUP_USER_IDENTITY -> freeUpUserIdentity

File: sunbird-platform-core/common-util/src/main/java/org/sunbird/common/request/UserFreeUpRequestValidator.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class UserFreeUpRequestValidator extends BaseRequestValidator {
+
+    private Request request;
+    private static List<String> identifiers = new ArrayList<>();
+    static {
+        identifiers.add("email");
```
Comment (manzarul) [inline sunbird-platform-core/common-util/src/main/java/org/sunbird/common/request/UserFreeUpRequestValidator.java]: use already define constant for email and phone

File: sunbird-platform-core/common-util/src/main/java/org/sunbird/common/request/UserFreeUpRequestValidator.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+                    ResponseCode.mandatoryParamsMissing.getErrorCode(),
+                    MessageFormat.format(ResponseCode.mandatoryParamsMissing.getErrorMessage(), JsonKey.IDENTIFIER),
+                    ResponseCode.CLIENT_ERROR.getResponseCode());
+        }
+
+    }
+
+    private void validateObject() {
+        Object identifierType = request.getRequest().get(JsonKey.IDENTIFIER);
+        if (!(identifierType instanceof List)) {
```
Comment (manzarul) [inline sunbird-platform-core/common-util/src/main/java/org/sunbird/common/request/UserFreeUpRequestValidator.java]: what will happen if identifierType is null

File: sunbird-platform-core/common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java [hunk 4]
```diff
@@ -437,6 +437,7 @@
     String ERROR_USER_UPDATE_PASSWORD = "User is created but password couldn't be updated.";
     String ERROR_BGMS_MISMATCH = "Mismatch in {0} at row - {1}";
     String ERROR_USER_MIGRATION_FAILED = "User migration failed.";
+    String IDENTIFIER_VALIDATION_FAILED="Valid identifier is not present in List, Valid supported identifiers are ";
```
Comment (manzarul) [inline sunbird-platform-core/common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: we already have same kind of list validation , can we not reused it.

File: sunbird-platform-core/common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java [hunk 5]
```diff
@@ -810,5 +811,6 @@
     String INVALID_REQUEST_TIMEOUT = "INVALID_REQUEST_TIMEOUT";
     String ERROR_BGMS_MISMATCH = "ERROR_BGMS_MISMATCH";
     String ERROR_USER_MIGRATION_FAILED = "ERROR_USER_MIGRATION_FAILED";
+    String VALID_IDENTIFIER_ABSENSE=" IDENTIFIER IN LIST IS NOT SUPPORTED OR INCORRECT";
```
Comment (manzarul) [inline sunbird-platform-core/common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: trim  space.

File: sunbird-platform-core/common-util/src/main/java/org/sunbird/common/request/UserFreeUpRequestValidator.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+import org.sunbird.common.models.util.JsonKey;
+import org.sunbird.common.models.util.ProjectUtil;
+import org.sunbird.common.responsecode.ResponseCode;
+
+import java.text.MessageFormat;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class UserFreeUpRequestValidator extends BaseRequestValidator {
```
Comment (manzarul) [inline sunbird-platform-core/common-util/src/main/java/org/sunbird/common/request/UserFreeUpRequestValidator.java]: All class and public method should have java doc.

File: sunbird-platform-core/common-util/src/main/java/org/sunbird/common/request/UserFreeUpRequestValidator.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    private void validateIdentifier() {
+        validatePresence();
+        validateObject();
+        validateSubset();
+    }
+
+
+    private void validatePresence() {
```
Comment (anmol2302) [inline sunbird-platform-core/common-util/src/main/java/org/sunbird/common/request/UserFreeUpRequestValidator.java]: i am validating it here

File: sunbird-platform-core/common-util/src/main/java/org/sunbird/common/request/UserFreeUpRequestValidator.java (hunk 2 above)
Comment (anmol2302) [reply sunbird-platform-core/common-util/src/main/java/org/sunbird/common/request/UserFreeUpRequestValidator.java]: done

File: sunbird-platform-core/common-util/src/main/java/org/sunbird/common/request/UserFreeUpRequestValidator.java (hunk 3 above)
Comment (anmol2302) [reply sunbird-platform-core/common-util/src/main/java/org/sunbird/common/request/UserFreeUpRequestValidator.java]:
```{
    "request": {
        "id": "061bc053-f744-487e-a060-ae91456bf2b9",
        "identifier":null
    }
}```
Response will be 
```{
    "id": "",
    "ver": "private",
    "ts": "2019-07-26 14:00:25:480+0530",
    "params": {
        "resmsgid": null,
        "msgid": "d219fca5-cd8d-4457-a679-1d3bee090acb",
        "err": "DATA_TYPE_ERROR",
        "status": "DATA_TYPE_ERROR",
        "errmsg": "Data type of identifier should be List."
    },
    "responseCode": "CLIENT_ERROR",
    "result": {}
}```

File: sunbird-platform-core/common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java (hunk 5 above)
Comment (anmol2302) [reply sunbird-platform-core/common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: done

File: sunbird-platform-core/common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java (hunk 4 above)
Comment (anmol2302) [reply sunbird-platform-core/common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java]: No, they are different sir

File: sunbird-platform-core/common-util/src/main/java/org/sunbird/common/request/UserFreeUpRequestValidator.java (hunk 6 above)
Comment (anmol2302) [reply sunbird-platform-core/common-util/src/main/java/org/sunbird/common/request/UserFreeUpRequestValidator.java]: done

File: sunbird-platform-core/common-util/src/main/java/org/sunbird/common/models/util/ActorOperations.java (hunk 1 above)
Comment (anmol2302) [reply sunbird-platform-core/common-util/src/main/java/org/sunbird/common/models/util/ActorOperations.java]: done

**Pull request merged**